### PR TITLE
Scatter Node

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

File: src/main/scala/wdl4s/wom/graph/GraphNode.scala [hunk 1]
```diff
@@ -47,4 +49,37 @@ object GraphNode {
 
     required.toList.traverse(requiredInputFound).void
   }
+
+  private[graph] class GraphNodeSetter {
+    var _graphNode: GraphNode = _
+    private def getGraphNode = _graphNode
+    def get: () => GraphNode = () => getGraphNode
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/GraphNode.scala]: should be `Unit => GraphNode` for the type

File: src/main/scala/wdl4s/wom/graph/GraphNode.scala [hunk 2]
```diff
@@ -47,4 +49,37 @@ object GraphNode {
...
+  }
+
+  private[graph] def linkInputPort(inputPrefix: String, inputMapping: Map[String, OutputPort], callNodeRef: () => GraphNode)(inputDefinition: Callable.InputDefinition): (InputPort, Option[GraphNode]) = {
+    val (outputPort, graphNode): (OutputPort, Option[GraphNode]) = inputDefinition match {
+      case input if inputMapping.contains(input.name) => (inputMapping(input.name), None)
+      case RequiredInputDefinition(n, womType) =>
+        val result = RequiredGraphInputNode(s"$inputPrefix$n", womType)
+        (result.singleOutputPort, Some(result))
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/GraphNode.scala]: `Option`s instead of `Some`s

File: src/main/scala/wdl4s/wom/graph/GraphNode.scala [hunk 3]
```diff
@@ -47,4 +49,37 @@ object GraphNode {
...
     required.toList.traverse(requiredInputFound).void
   }
+
+  private[graph] class GraphNodeSetter {
+    var _graphNode: GraphNode = _
+    private def getGraphNode = _graphNode
+    def get: () => GraphNode = () => getGraphNode
+  }
+
+  private[graph] def linkInputPort(inputPrefix: String, inputMapping: Map[String, OutputPort], callNodeRef: () => GraphNode)(inputDefinition: Callable.InputDefinition): (InputPort, Option[GraphNode]) = {
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/GraphNode.scala]: type `Unit`

File: src/main/scala/wdl4s/wom/graph/GraphNodePort.scala [hunk 4]
```diff
@@ -25,37 +25,22 @@ object GraphNodePort {
     // TODO: Might end up wanting a backwards link to the InputPorts that use this (eg def downstream: Set[InputPort])?
   }
 
-  final case class ConnectedInputPort(name: String, womType: WdlType, upstreamPort: OutputPort, g: () => GraphNode) extends InputPort {
-    override def upstream = upstreamPort
-    override lazy val graphNode: GraphNode = g()
+  sealed trait DelayedGraphNodePort { this: GraphNodePort =>
+    val g: () => GraphNode
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/GraphNodePort.scala]: type `Unit`

File: src/main/scala/wdl4s/wom/graph/GraphNodePort.scala [hunk 5]
```diff
@@ -25,37 +25,22 @@ object GraphNodePort {
...
     */
-  final case class DeclarationOutputPort(name: String, womType: WdlType, graphNode: GraphNode) extends OutputPort
+  final case class GraphNodeOutputPort(name: String, womType: WdlType, graphNode: GraphNode) extends OutputPort
 
   // TODO: For these next two, the graphNode should be a ScatterNode and IfNode respectively (once those exist):
   /**
     * Represents the gathered output from a call/declaration in a scatter.
     */
-  final case class ScatterGathererPort(name: String, womType: WdlArrayType, outputToGather: OutputPort, graphNode: GraphNode) extends OutputPort
+  final case class ScatterGathererPort(name: String, womType: WdlArrayType, outputToGather: GraphOutputNode, g: () => GraphNode) extends OutputPort with DelayedGraphNodePort
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/GraphNodePort.scala]: ditto

File: src/main/scala/wdl4s/wom/graph/hmm.wdl [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  call a
+  call b
+
+  scatter(x in a.out) {
+    call c { input: i = b.out }
+  }
+
+  call d { input: i = c.out }
+
+}
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/hmm.wdl]: \n

File: src/main/scala/wdl4s/wom/graph/GraphNode.scala [hunk 7]
```diff
@@ -47,4 +49,37 @@ object GraphNode {
 
     required.toList.traverse(requiredInputFound).void
   }
+
+  private[graph] class GraphNodeSetter {
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/GraphNode.scala]: what is this?

Comment (danbills) [inline src/main/scala/wdl4s/wom/graph/GraphNode.scala]: This looks like a concurrent promise?  Maybe a comment regarding the approach?

File: src/main/scala/wdl4s/wom/graph/GraphNode.scala (hunk 3 above)
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/GraphNode.scala]: the return type feels like a type and by type I mean `case class`

File: src/main/scala/wdl4s/wom/graph/hmm.wdl [hunk 8]
```diff
@@ -0,0 +1,11 @@
+workflow foo {
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/hmm.wdl]: the name of this file makes me go hmm in more ways than one

File: src/main/scala/wdl4s/wom/graph/ScatterNode.scala [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    * @param innerGraph The inner graph that's being scattered over.
+    * @param scatterVariableSource The OutputPort (in the outer Graph) to connect to the ScatterNode's scatterVariable InputPort.
+    * @param scatterVariableInput the GraphInputNode in the innerGraph which represents the scatter variable.
+    * @param inputMapping The mapping to inputs expected by the innerGraph
+    * @return A 2-tuple of:
+    *         * _1: The scatter node itself, connected to the inputs
+    *         * _2: A set of GraphInputNodes for inputs which weren't supplied
+    */
+  // TODO: Validate that the scatter source is an array type (oh how I wish ports could be typed!)
+  def scatterOverGraph(innerGraph: Graph, scatterVariableSource: OutputPort, scatterVariableInput: GraphInputNode, inputMapping: Map[String, OutputPort]): (ScatterNode, Set[GraphNode]) = {
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/ScatterNode.scala]: another `case class` return type?

File: src/main/scala/wdl4s/wom/graph/CallNode.scala [hunk 10]
```diff
@@ -62,25 +56,9 @@ object CallNode {
...
-      }
-
-      (ConnectedInputPort(inputDefinition.name, inputDefinition.womType, outputPort, callNodeRef), graphNode)
-    }
-
-    val graphNodeSetter = new GraphNodeSetter()
-    val linkedInputPortsAndGraphInputNodes: Set[(InputPort, Option[GraphNode])] = callable.inputs.map(linkInputPort(_, graphNodeSetter.get))
+    val graphNodeSetter = new GraphNode.GraphNodeSetter()
+    val inputPortLinker = GraphNode.linkInputPort(callable.name + ".", inputMapping, graphNodeSetter.get) _
+    val linkedInputPortsAndGraphInputNodes: Set[(InputPort, Option[GraphNode])] = callable.inputs.map(inputPortLinker)
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/CallNode.scala]: ToL if it was me `callable.inputs map inputPortLinker`

File: src/main/scala/wdl4s/wom/graph/GraphNode.scala (hunk 7 above)
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/GraphNode.scala]: discussed in person

File: src/main/scala/wdl4s/wom/graph/GraphNode.scala (hunk 3 above)
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/GraphNode.scala]: some high level comments as to what this does would be awesome

File: src/main/scala/wdl4s/wom/graph/GraphNode.scala [hunk 11]
```diff
@@ -47,4 +49,37 @@ object GraphNode {
...
+        (result.singleOutputPort, Some(result))
+      case OptionalInputDefinitionWithDefault(n, womType, default) =>
+        val result = OptionalGraphInputNodeWithDefault(s"$inputPrefix$n", womType, default)
+        (result.singleOutputPort, Some(result))
+    }
+
+    (ConnectedInputPort(inputDefinition.name, inputDefinition.womType, outputPort, callNodeRef), graphNode)
+  }
+
+  private[wom] implicit class EnhancedGraphNodeSet(nodes: Set[GraphNode]) {
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/GraphNode.scala]: `extends AnyVal`

File: src/main/scala/wdl4s/wom/graph/GraphNodePort.scala [hunk 12]
```diff
@@ -25,37 +25,22 @@ object GraphNodePort {
     // TODO: Might end up wanting a backwards link to the InputPorts that use this (eg def downstream: Set[InputPort])?
   }
 
-  final case class ConnectedInputPort(name: String, womType: WdlType, upstreamPort: OutputPort, g: () => GraphNode) extends InputPort {
-    override def upstream = upstreamPort
-    override lazy val graphNode: GraphNode = g()
+  sealed trait DelayedGraphNodePort { this: GraphNodePort =>
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/GraphNodePort.scala]: comment plz

File: src/main/scala/wdl4s/wom/graph/GraphNodePort.scala (hunk 4 above)
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/GraphNodePort.scala]: what up (with this) g?

File: src/main/scala/wdl4s/wom/graph/ScatterNode.scala [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+final case class ScatterNode private(graph: Graph,
+                                     scatterVariableMapping: ScatterVariableMapping,
+                                     otherInputPorts: Set[InputPort],
+                                     outputMapping: Set[ScatterGathererPort]) extends GraphNode {
+
+  def innerGraphInputs: Set[GraphInputNode] = graph.nodes.collect { case gin: GraphInputNode => gin }
+
+  // NB if you find yourself calling .filter on this set of inputPorts, you probably just wanted to access either
+  // the scatterVariableMapping or otherInputPorts fields directly.
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/ScatterNode.scala]: what exactly falls under the heading of "other" in `otherInputPorts`?

File: src/main/scala/wdl4s/wom/graph/ScatterNode.scala [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+  override val inputPorts: Set[InputPort] = Set(scatterVariableMapping.inputPort).union(otherInputPorts)
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/ScatterNode.scala]: why not `Set(...) ++ otherInputPorts`?

File: src/main/scala/wdl4s/wom/graph/ScatterNode.scala [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    */
+  case class ScatterVariableMapping(inputPort: InputPort, graphInputNode: GraphInputNode)
+
+  /**
+    * Creates a ScatterNode that's been connected to the inputs provided.
+    * @param innerGraph The inner graph that's being scattered over.
+    * @param scatterVariableSource The OutputPort (in the outer Graph) to connect to the ScatterNode's scatterVariable InputPort.
+    * @param scatterVariableInput the GraphInputNode in the innerGraph which represents the scatter variable.
+    * @param inputMapping The mapping to inputs expected by the innerGraph
+    * @return A 2-tuple of:
```
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/ScatterNode.scala]: this particular case is definitely pronounced "twople"

File: src/main/scala/wdl4s/wom/graph/hmm.wdl (hunk 8 above)
Comment (cjllanwarne) [inline src/main/scala/wdl4s/wom/graph/hmm.wdl]: this looks a lot like a "I'll just test some WDLing in intellij for the syntax highlighting" which accidentally got committed

File: src/main/scala/wdl4s/wom/graph/ScatterNode.scala (hunk 13 above)
Comment (cjllanwarne) [inline src/main/scala/wdl4s/wom/graph/ScatterNode.scala]: Imagine that the `scatter` block were a self-contained unit and you had to write an inputs JSON for it... It's **that** set minus the scatter variable.

File: src/main/scala/wdl4s/wom/graph/ScatterNode.scala (hunk 14 above)
Comment (cjllanwarne) [inline src/main/scala/wdl4s/wom/graph/ScatterNode.scala]: I didn't find a `++` on `Set`. And union seemed... appropriate?

File: src/test/scala/wdl4s/wom/graph/ScatterNodeSpec.scala [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+
+    val x_inputNode = RequiredGraphInputNode("x", WdlIntegerType)
+    val (foo_callNode, _) = CallNode.callWithInputs("foo", task_foo, Map("i" -> x_inputNode.singleOutputPort))
+    val scatterGraph = Graph.validateAndConstruct(Set(foo_callNode, x_inputNode)) match {
+      case Valid(sg) => sg.withDefaultOutputs
+      case Invalid(es) => fail("Failed to make scatter graph: " + es.toList.mkString(", "))
+    }
+
+    val (scatterNode, unsuppliedScatterInputNodes) = ScatterNode.scatterOverGraph(scatterGraph, xs_inputNode.singleOutputPort, x_inputNode, Map.empty)
+    unsuppliedScatterInputNodes.size should be(0)
```
Comment (mcovarr) [inline src/test/scala/wdl4s/wom/graph/ScatterNodeSpec.scala]: there's probably a `shouldBe empty` or something

File: src/main/scala/wdl4s/wom/graph/ScatterNode.scala (hunk 14 above)
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/ScatterNode.scala]: it's there 😄

File: src/main/scala/wdl4s/wom/graph/ScatterNode.scala (hunk 13 above)
Comment (mcovarr) [inline src/main/scala/wdl4s/wom/graph/ScatterNode.scala]: That explanation immortalized in a comment would be lovely.

**Comment by mcovarr** [comment]
:+1: my biggest request would be high-level comments.  As I haven't been deep in WOM yet but expect to be soon, it's taking me a while to grok what's going on.

File: src/main/scala/wdl4s/wom/graph/ScatterNode.scala (hunk 15 above)
Comment (curoli) [inline src/main/scala/wdl4s/wom/graph/ScatterNode.scala]: I'd think it's "double" or "pair".

**Comment by danbills** [comment]
👍

File: src/main/scala/wdl4s/wom/graph/ScatterNode.scala (hunk 13 above)
Comment (cjllanwarne) [inline src/main/scala/wdl4s/wom/graph/ScatterNode.scala]: Done

**Pull request merged**